numpy/_core/tests/test_stringdtype.py

```diff
@@ -0,0 +1,1159 @@
+    with pytest.raises(OverflowError):
+        np.array(oob, dtype="T").astype(idtype)
+
+
+@pytest.mark.parametrize("typename", ["byte", "short", "int", "longlong"])
+@pytest.mark.parametrize("signed", ["", "u"])
+def test_unsized_integer_casts(typename, signed):
+    idtype = f"{signed}{typename}"
+
+    inp = [1, 2, 3, 4]
+    ainp = np.array(inp, dtype=idtype)
+    assert_array_equal(ainp, ainp.astype("T").astype(idtype))
+
+
+@pytest.mark.parametrize(
+    "typename",
+    [
+        pytest.param(
+            "longdouble",
+            marks=pytest.mark.xfail(
+                np.dtypes.LongDoubleDType() != np.dtypes.Float64DType(),
+                reason="numpy lacks an ld2a implementation",
+                strict=True,
+            ),
+        ),
+        "float64",
+        "float32",
+        "float16",
+    ],
+)
+def test_float_casts(typename):
+    inp = [1.1, 2.8, -3.2, 2.7e4]
+    ainp = np.array(inp, dtype=typename)
+    assert_array_equal(ainp, ainp.astype("T").astype(typename))
+
+    inp = [0.1]
+    sres = np.array(inp, dtype=typename).astype("T")
+    res = sres.astype(typename)
+    assert_array_equal(np.array(inp, dtype=typename), res)
+    assert sres[0] == "0.1"
+
+    if typename == "longdouble":
+        # let's not worry about platform-dependent rounding of longdouble
+        return
+
+    fi = np.finfo(typename)
+
+    inp = [1e-324, fi.smallest_subnormal, -1e-324, -fi.smallest_subnormal]
+    eres = [0, fi.smallest_subnormal, -0, -fi.smallest_subnormal]
+    res = np.array(inp, dtype=typename).astype("T").astype(typename)
+    assert_array_equal(eres, res)
+
+    inp = [2e308, fi.max, -2e308, fi.min]
+    eres = [np.inf, fi.max, -np.inf, fi.min]
+    res = np.array(inp, dtype=typename).astype("T").astype(typename)
+    assert_array_equal(eres, res)
+
+
+@pytest.mark.parametrize(
+    "typename",
+    [
+        "csingle",
+        "cdouble",
+        pytest.param(
+            "clongdouble",
+            marks=pytest.mark.xfail(
+                np.dtypes.CLongDoubleDType() != np.dtypes.Complex128DType(),
+                reason="numpy lacks an ld2a implementation",
+                strict=True,
+            ),
+        ),
+    ],
+)
+def test_cfloat_casts(typename):
+    inp = [1.1 + 1.1j, 2.8 + 2.8j, -3.2 - 3.2j, 2.7e4 + 2.7e4j]
+    ainp = np.array(inp, dtype=typename)
+    assert_array_equal(ainp, ainp.astype("T").astype(typename))
+
+    inp = [0.1 + 0.1j]
+    sres = np.array(inp, dtype=typename).astype("T")
+    res = sres.astype(typename)
+    assert_array_equal(np.array(inp, dtype=typename), res)
+    assert sres[0] == "(0.1+0.1j)"
+
+
+def test_take(string_list):
+    sarr = np.array(string_list, dtype="T")
+    res = sarr.take(np.arange(len(string_list)))
+    assert_array_equal(sarr, res)
+
+    # make sure it also works for out
+    out = np.empty(len(string_list), dtype="T")
+    out[0] = "hello"
+    res = sarr.take(np.arange(len(string_list)), out=out)
+    assert res is out
+
```

Comment:
needs a repeat of `assert_array_equal(sarr, res)` just to be sure the "hello" is overwritten...